docs/specs/fee-calculator.md

```diff
@@ -0,0 +1,379 @@
+3. **Data Gaps**: Throw error if critical data is missing
+4. **Zero Balances**: Valid scenario, process normally
+
+## 7. Data Structures
+
+### Input Types
+
+```typescript
+interface DistributorsData {
+  distributors: Array<{
+    address: string;
+    metadata: {
+      name: string;
+      created_at: string;
+    };
+  }>;
+}
+
+interface BalanceData {
+  metadata: {
+    chain_id: number;
+    reward_distributor: string;
+  };
+  balances: {
+    [date: string]: {
+      block_number: number;
+      balance_wei: string;
+    };
+  };
+}
+
+interface RecipientRecievedEventData {
+  metadata: {
+    chain_id: number;
+    distributor_address: string;
+  };
+  events: Array<{
+    block_number: number;
+    transaction_hash: string;
+    log_index: number;
+    timestamp: string;
+    date: string;
+    recipient: string;
+    amount_wei: string;
+  }>;
+}
+```
+
+### Output Types
+
+```typescript
+interface FeeReport {
+  metadata: {
+    chain_id: number;
+  };
+  distributors: {
+    [distributorAddress: string]: {
+      name: string;
+      daily_totals: Array<{
+        date: string;
+        start_balance_wei: string;
+        end_balance_wei: string;
+        balance_change_wei: string;
+        distributions_wei: string;
+        distributions_count: number;
+        total_wei: string;
+      }>;
+    };
+  };
+}
+```
+
+## 8. Implementation Requirements
+
+### Validation Rules
+
+1. **Address Validation**:
+
+   - All addresses must be valid Ethereum addresses
+   - Addresses normalized to checksummed format
+
+2. **Date Validation**:
+
+   - All dates in ISO format (YYYY-MM-DD)
+   - Dates must be chronologically ordered
+
+3. **Amount Validation**:
+
+   - All amounts must be valid numeric strings
+   - Handle large numbers using bigint internally
+   - Convert back to strings for output
+
+4. **Data Completeness**:
+   - Throw error if critical data is missing
+
+### Security Considerations
+
+1. **FileManager ensures atomic file operations**
+
+2. **Error Handling**:
+   - Never expose system paths in error messages
+   - Fail gracefully with descriptive error messages
+
+## 9. Test Data Requirements
+
+### Minimum Test Data
+
+Integration tests require real blockchain data from Arbitrum Nova:
+
+1. **Distributors**: Use actual reward distributor addresses from test data:
+
+   - At least 5 distributors with varying creation dates
+   - Mix of high-volume and low-volume distributors
+   - Include distributors created at different times
+
+2. **Balance Data**: For each test distributor:
+
+   - Complete balance history for test period (minimum 30 days)
+   - Include days with increasing and decreasing balances
+   - Cover edge cases: zero balances, large balances
+
+3. **Distribution Events**: For each test distributor:
+
+   - All distribution events during test period
+   - Include days with multiple distributions
+   - Include days with no distributions
+   - Varying distribution amounts
+
+4. **Test Scenarios**:
```

Comment:
-let's just write that the implementer of the integration tests should be able to choose what tests they want to write. It's up to their discretion basically, so you can get rid of the specific tests here.